File: app/src/main/java/com/example/videoplayer/ui/MainActivity.java
package com.example.videoplayer.ui;

import android.content.Intent;
import android.os.Bundle;
import android.view.animation.AnimationUtils;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityOptionsCompat;

import com.example.videoplayer.R;
import com.example.videoplayer.databinding.ActivityMainBinding;

/**
 * Entry point – a URL-input screen that lets the user enter any M3U8 URL
 * and launch the player. Ships with a sample HLS stream URL pre-filled.
 */
public class MainActivity extends AppCompatActivity {

    private ActivityMainBinding binding;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivityMainBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        setupAnimations();
        setupClickListeners();
    }

    private void setupAnimations() {
        try {
            android.view.animation.Animation slideUp = AnimationUtils.loadAnimation(this, R.anim.slide_up_fade_in);
            binding.cardPlayer.startAnimation(slideUp);
        } catch (Exception ignored) {}
    }

    private void setupClickListeners() {
        binding.btnPlay.setOnClickListener(v -> {
            String url = binding.etStreamUrl.getText() != null
                    ? binding.etStreamUrl.getText().toString().trim()
                    : null;
            String tracksJson = binding.etTracksJson.getText() != null
                    ? binding.etTracksJson.getText().toString().trim()
                    : "";
            String videoUrlApi = binding.etVideoUrlApi.getText() != null
                    ? binding.etVideoUrlApi.getText().toString().trim()
                    : "";

            if (url == null || url.isEmpty()) {
                binding.tilStreamUrl.setError("Please enter a valid M3U8 URL");
                return;
            }
            binding.tilStreamUrl.setError(null);

            launchPlayer(
                    url,
                    tracksJson.isEmpty() ? null : tracksJson,
                    videoUrlApi.isEmpty() ? null : videoUrlApi
            );
        });
    }

    private void launchPlayer(String url, String tracksJson, String videoUrlApi) {
        android.util.Log.i("MainActivity",
                "Launching player url=" + url +
                " tracksJsonPresent=" + (tracksJson != null && !tracksJson.isEmpty()) +
                " videoUrlApiPresent=" + (videoUrlApi != null && !videoUrlApi.isEmpty()));

        Intent intent = new Intent(this, PlayerActivity.class);
        intent.putExtra(PlayerActivity.EXTRA_STREAM_URL, url);
        if (tracksJson != null && !tracksJson.isEmpty()) {
            intent.putExtra(PlayerActivity.EXTRA_TRACKS_PAYLOAD, tracksJson);
        }
        if (videoUrlApi != null && !videoUrlApi.isEmpty()) {
            intent.putExtra(PlayerActivity.EXTRA_VIDEO_URL_API, videoUrlApi);
        }

        Bundle options = ActivityOptionsCompat.makeCustomAnimation(
                this, R.anim.slide_in_right, R.anim.slide_out_left
        ).toBundle();
        startActivity(intent, options);
    }
}

File: app/src/main/java/com/example/videoplayer/ui/MainActivity.kt
package com.example.videoplayer.ui

import android.content.Intent
import android.os.Bundle
import androidx.core.app.ActivityOptionsCompat
import android.view.animation.AnimationUtils
import androidx.appcompat.app.AppCompatActivity
import com.example.videoplayer.R
import com.example.videoplayer.databinding.ActivityMainBinding

/**
 * Entry point – a URL-input screen that lets the user enter any M3U8 URL
 * and launch the player. Ships with a sample HLS stream URL pre-filled.
 *
 * Fields:
 *  - Stream URL      : The HLS .m3u8 URL to play initially.
 *  - Tracks JSON     : The 'tracks' JSON array from the movie details API response.
 *  - Video URL API   : The video-URL fetch API endpoint. Required for tracks
 *                      with existIndividualVideo=true. The player appends
 *                      &languageId=<id> to this URL when user selects such a track.
 */
class MainActivity : AppCompatActivity() {

    private lateinit var binding: ActivityMainBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        setupAnimations()
        setupClickListeners()
    }

    private fun setupAnimations() {
        val slideUp = AnimationUtils.loadAnimation(this, R.anim.slide_up_fade_in)
        binding.cardPlayer.startAnimation(slideUp)
    }

    private fun setupClickListeners() {
        binding.btnPlay.setOnClickListener {
            val url = binding.etStreamUrl.text?.toString()?.trim()
            val tracksJson = binding.etTracksJson.text?.toString()?.trim().orEmpty()
            val videoUrlApi = binding.etVideoUrlApi.text?.toString()?.trim().orEmpty()

            if (url.isNullOrEmpty()) {
                binding.tilStreamUrl.error = "Please enter a valid M3U8 URL"
                return@setOnClickListener
            }
            binding.tilStreamUrl.error = null

            launchPlayer(
                url = url,
                tracksJson = tracksJson.ifBlank { null },
                videoUrlApi = videoUrlApi.ifBlank { null }
            )
        }
    }

    private fun launchPlayer(
        url: String,
        tracksJson: String? = null,
        videoUrlApi: String? = null
    ) {
        android.util.Log.i(
            "MainActivity",
            "Launching player url=$url " +
                "tracksJsonPresent=${!tracksJson.isNullOrBlank()} " +
                "videoUrlApiPresent=${!videoUrlApi.isNullOrBlank()}"
        )
        val intent = Intent(this, PlayerActivity::class.java).apply {
            putExtra(PlayerActivity.EXTRA_STREAM_URL, url)
            if (!tracksJson.isNullOrBlank()) {
                putExtra(PlayerActivity.EXTRA_TRACKS_PAYLOAD, tracksJson)
            }
            if (!videoUrlApi.isNullOrBlank()) {
                putExtra(PlayerActivity.EXTRA_VIDEO_URL_API, videoUrlApi)
            }
        }
        val options = ActivityOptionsCompat.makeCustomAnimation(
            this, R.anim.slide_in_right, R.anim.slide_out_left
        )
        startActivity(intent, options.toBundle())
    }
}
